### PR TITLE
Create Input, Fieldset, and Field components for terra-form

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
Our last forms pull request was massive 😮 .

Let's make things a littler easier, by having small plates at a time!

For this review, we will start with the standard Field, Fieldset, and Input elements. Those will be the base for other individual terra components.

### Additional Details
Once this review is done, I'll start working on the Control component next as that will have more complexity, followed by TextField, Numberfield, etc.

### CODE REVIEW

File: packages/terra-form/src/components/Field.scss [hunk 1]
```diff
@@ -0,0 +1,55 @@
+@import '~terra-mixins';
+@import '../_variables.scss';
```
Comment (bjankord) [inline packages/terra-form/src/components/Field.scss]: When importing a partial .scss file, you can omit the underscore prefix.
e.g. `@import '../variables.scss';`

File: packages/terra-form/src/components/Input.scss [hunk 2]
```diff
@@ -0,0 +1,18 @@
+@import '~terra-mixins';
+@import '../_variables.scss';
```
Comment (bjankord) [inline packages/terra-form/src/components/Input.scss]: When importing a partial .scss file, you can omit the underscore prefix.
e.g. `@import '../variables.scss';`

File: packages/terra-form/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Terra Form
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-form.svg)](https://www.npmjs.org/package/terra-form)
+[![Build Status](https://travis-ci.org/cerner/terra-core.svg?branch=master)](https://travis-ci.org/cerner/terra-core)
+
+{insert description}
```
Comment (emilyrohrbough) [inline packages/terra-form/README.md]: Add description

File: packages/terra-form/docs/themeable-variables.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Themeable Variables
+
+Variables to `theme` the terra-form component.
+
+| Variable Name                  | Default value            | Usage                                     |
+|--------------------------------|--------------------------|-------------------------------------------|
+| $terra-form-required-color     | $terra-red-70            | Color used to denote a field as required. |
+| $terra-form-required-indicator | *                        | Indicator to use for required fields.     |
```
Comment (emilyrohrbough) [inline packages/terra-form/docs/themeable-variables.md]: Is the $terra-form-required-indicator a themeable string value?

File: packages/terra-form/lib/_variables.scss [hunk 5]
```diff
@@ -0,0 +1,8 @@
+$terra-form-required-color: $terra-red-70;
+$terra-form-required-indicator: '*';
+$terra-form-error-color: $terra-red-70;
+$terra-form-label-font-weight: bold;
+$terra-form-legend-font-weight: bold;
+$terra-form-field-border: 0;
+$terra-form-legend-color: $terra-dark;
+$terra-form-input-border: $terra-grey-15 solid 1px;
```
Comment (emilyrohrbough) [inline packages/terra-form/lib/_variables.scss]: @neilpfeiffer Did we decided if themeable variables like border is themeable by color or also by border width, etc?

File: packages/terra-form/package.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "terra-form",
+  "main": "lib/Form.js",
+  "private": true,
+  "version": "0.0.0",
+  "description": "terra-form",
```
Comment (emilyrohrbough) [inline packages/terra-form/package.json]: Add description

File: packages/terra-form/package.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "prop-types": "^15.5.8",
+    "terra-base": "^0.x",
+    "terra-mixins": "^1.0.0"
+  },
+  "scripts": {
+    "compile": "npm run compile:clean && npm run compile:build",
+    "compile:clean": "rm -rf lib",
```
Comment (emilyrohrbough) [inline packages/terra-form/package.json]: Update to use rimraf 
`"compile:clean": "$(cd ..; npm bin)/rimraf lib",`

File: packages/terra-form/src/components/Field.jsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    customProps.className,
+  );
+
+  return (
+    <p {...customProps} className={fieldClasses}>
+      {label && <label className="terra-Form-label">{label}</label>}
+      {children}
+      {help && <small className="terra-Form-helpText">{help}</small>}
+      {error && <small className="terra-Form-error" tabIndex="-1">{error}</small>}
+    </p>
```
Comment (emilyrohrbough) [inline packages/terra-form/src/components/Field.jsx]: suggestion from old PR: https://github.com/cerner/terra-core/pull/305/files#diff-89873514d1b4737a5dd30a8c352d0088R53

File: packages/terra-form/src/components/Fieldset.jsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  error: null,
+  help: null,
+  htmlFor: null,
+  isInline: false,
+  legend: null,
+};
+
+const Fieldset = ({ children, error, help, htmlFor, isInline, legend, ...customProps }) => {
+  const fieldsetClasses = classNames(
+    'terra-Form-field',
```
Comment (emilyrohrbough) [inline packages/terra-form/src/components/Fieldset.jsx]: Would we want a specific css class for fieldset to indicate the difference between a field component and a fieldset component?

File: packages/terra-form/tests/jest/components/Input.test.jsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  const input = (
+    <Input
+      required
+      name="foo"
+      value="bar"
+    />);
+
+  const wrapper = shallow(input);
+  expect(wrapper).toMatchSnapshot();
+});
```
Comment (emilyrohrbough) [inline packages/terra-form/tests/jest/components/Input.test.jsx]: Is there value in adding snapshot tests for the differences between controlled and uncontrolled components?

File: packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  },
+
+  'Displays a populated Field with correct label, help message, and choices': (browser) => {
+    browser
+      .url(`http://localhost:${browser.globals.webpackDevServerPort}/#/tests/form-tests/fieldset/populated`)
+      .assert.elementPresent('.terra-Form-field')
+      .assert.elementPresent('.terra-Form-field--inline')
+      .assert.containsText('.terra-Form-legend', 'Do you have any Children?')
+      .assert.containsText('.terra-Form-helpText', 'Families are eligible for family package plans')
+      .assert.containsText('.terra-Form-error', 'This field is required');
```
Comment (emilyrohrbough) [inline packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js]: How are these error messages being displayed?

File: packages/terra-site/src/examples/form/examples/Input.jsx [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+const InputExamples = () => (
+  <form>
+    <Input
+      name="foo"
+      defaultValue="bar"
+    />
+  </form>
+);
+
+export default InputExamples;
```
Comment (emilyrohrbough) [inline packages/terra-site/src/examples/form/examples/Input.jsx]: Should we include both a controlled & uncontrolled example?

**Comment by emilyrohrbough** [comment]
Missing jest tests for Field and Fieldset.

**Comment by JakeLaCombe** [comment]
Wow! I had them on another branch, and it looks like I forgot to copy them over. https://github.com/cerner/terra-core/pull/352/commits/9add5acb750961b3d5acda03fe56aa47534644e8

File: packages/terra-site/src/examples/form/examples/Input.jsx (hunk 12 above)
Comment (JakeLaCombe) [inline packages/terra-site/src/examples/form/examples/Input.jsx]: My plan is to add a form example at the bottom of the page to handle that for all components. How does that sound?

File: packages/terra-form/tests/jest/components/Input.test.jsx (hunk 10 above)
Comment (JakeLaCombe) [inline packages/terra-form/tests/jest/components/Input.test.jsx]: @emilyrohrbough See how these work for you. From looking at the jest docs, I think this is the best I can get, which I think is ok for our case.

https://github.com/cerner/terra-core/pull/352/commits/f57bf091b3c5985efe503336c3b5fb4c5ae50896

File: packages/terra-form/src/components/Fieldset.jsx (hunk 9 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Fieldset.jsx]: Right now in blue_steel, we keep them using the same. I debated on this for awhile with myself. http://snapshot.bluesteel.hax/forms/layouts

However, I think for future maintainability, it's better to have them separate

File: packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js (hunk 11 above)
Comment (JakeLaCombe) [inline packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js]: screen shot 2017-05-16 at 1 21 11 pm

We have fieldset errors looking like this atm.

File: packages/terra-form/src/components/Field.jsx (hunk 8 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Field.jsx]: https://github.com/cerner/terra-core/pull/352/commits/68361e531d04900062a343d7b777983c79df4334

File: packages/terra-form/src/components/Input.jsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  value: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.number,
+  ]),
+};
+
+const defaultProps = {
+  defaultValue: undefined,
+  name: null,
+  onChange: () => {},
```
Comment (Matt-Butler) [inline packages/terra-form/src/components/Input.jsx]: I think this should be removed.  If onChange is not provided, then no onChange function should be added to the input.

File: packages/terra-form/tests/jest/components/Input.test.jsx (hunk 10 above)
Comment (emilyrohrbough) [inline packages/terra-form/tests/jest/components/Input.test.jsx]: @JakeLaCombe Looks good. Did the snapshot file update form this?

File: packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js (hunk 11 above)
Comment (emilyrohrbough) [inline packages/terra-form/tests/nightwatch/components/fieldset/fieldset-spec.js]: Okay. Wasn't sure if the error message was rendered after the user supplied incorrect input.

File: packages/terra-site/package.json [hunk 14]
```diff
@@ -41,6 +41,7 @@
     "terra-content-container": "^0.x",
     "terra-date-picker": "^0.x",
     "terra-demographics-banner": "^0.x",
+    "terra-form": "^0.x.0",
```
Comment (Matt-Butler) [inline packages/terra-site/package.json]:
```
"^0.x"
```

**Comment by Matt-Butler** [comment]
I am seeing this warning
```
Warning: Input is changing a controlled input of type text to be uncontrolled. Input elements should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
```

controlled_uncontrolled

**Comment by Matt-Butler** [comment]
focus
Should this element be focusable?

File: packages/terra-form/src/components/Input.jsx [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  value,
+  ...customProps
+}) => {
+  const additionalInputProps = Object.assign({}, customProps);
+
+  if (required) {
+    additionalInputProps['aria-required'] = 'true';
+  }
+
+  if (value) {
```
Comment (Matt-Butler) [inline packages/terra-form/src/components/Input.jsx]: `""` an empty string will evaluate to false.

https://developer.mozilla.org/en-US/docs/Glossary/Falsy

File: packages/terra-form/tests/jest/components/Input.test.jsx (hunk 10 above)
Comment (JakeLaCombe) [inline packages/terra-form/tests/jest/components/Input.test.jsx]: For this one, it didn't, as we didn't have new snapshots to check against.

File: packages/terra-form/src/components/Fieldset.jsx (hunk 9 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Fieldset.jsx]: Added the new class. https://github.com/cerner/terra-core/pull/352/commits/3474f23aa78e34bde9b8047d43f69b2696e8ce07

File: packages/terra-form/src/components/Input.jsx (hunk 15 above)
Comment (Matt-Butler) [inline packages/terra-form/src/components/Input.jsx]: I would update this to check if value is undefined.

File: packages/terra-form/src/components/Input.jsx (hunk 13 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Input.jsx]: I agree. The easiest way to do this is to remove onChange as a first class attribute from the component, as it comes default.

https://github.com/cerner/terra-core/pull/352/commits/41cbef501d514b9499cdb298b1576db3bdd7195a

File: packages/terra-form/src/components/Input.jsx (hunk 15 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Input.jsx]: That did the trick.

https://github.com/cerner/terra-core/pull/352/commits/9c897f626b62fba1f2b35c6d5d000b7d911d6030

**Comment by windse7en** [comment]
The heroku reviews app can't be created due to the existing app name: terra-core-deployed-pr-352.
I tried to access to that server, but I don't have the access. Someone created a heroku app with our pr-352 name.

To deploy to the heroku, my suggest:
1. We can create a heroku app from your form branch, normally deploy as heroku app.
2. Recreate another PR to avoid this name conflict.

**Comment by JakeLaCombe** [comment]
New deploy url! http://terra-forms.herokuapp.com/static/#/site/form

**Comment by JakeLaCombe** [comment]
@Matt-Butler The error messages have an interesting story behind them. I looked at the existing terra base, and found that all of the error and help messages had tabindex=-1 on them. This is due to issues with IE screen readers not being able to properly handle the aria-labelledby and aria-describedby attributes in the MSAA accessibility API. To fix this, microsoft provided a fix of attaching tabindex=-1 to the desired nodes. You can read more about it [here](https://www.paciellogroup.com/blog/2014/06/aria-labelledby-aria-describedby-support-popular-windows-browsers/).

To fix what you saw, I removed the outline from the error and help messages, and added tabindex=-1 to the help messages. https://github.com/cerner/terra-core/pull/352/commits/970e6a5f5b53c0672eec8f85511c5dc5f6cad3a1

File: packages/terra-site/package.json (hunk 14 above)
Comment (JakeLaCombe) [inline packages/terra-site/package.json]: https://github.com/cerner/terra-core/pull/352/commits/2e129c61e6972077c23e0053be803b11b886b0ca

File: packages/terra-form/package.json (hunk 7 above)
Comment (JakeLaCombe) [inline packages/terra-form/package.json]: https://github.com/cerner/terra-core/pull/352/commits/0aee37909843eb8fc0b95a20a9f2868c48763be3

File: packages/terra-form/package.json (hunk 6 above)
Comment (JakeLaCombe) [inline packages/terra-form/package.json]: https://github.com/cerner/terra-core/pull/352/commits/6242595b5b52cfa5f695be1f0e540a009c6f40de

File: packages/terra-form/README.md (hunk 3 above)
Comment (JakeLaCombe) [inline packages/terra-form/README.md]: https://github.com/cerner/terra-core/pull/352/commits/29d827a324b359afdb14063424ae363f20a44d3f

File: packages/terra-form/src/components/Input.scss (hunk 2 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Input.scss]: Fixed it here - https://github.com/cerner/terra-core/pull/352/commits/594716abdc84473dfcf3ced97dc9fb5897d5666c

File: packages/terra-form/src/components/Field.scss (hunk 1 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Field.scss]: Fixed it here - https://github.com/cerner/terra-core/pull/352/commits/594716abdc84473dfcf3ced97dc9fb5897d5666c

File: packages/terra-form/docs/components/Field.md [hunk 16]
```diff
@@ -0,0 +1,23 @@
+# Field
+
+Generic form field component which handles the layout of a standard form field including errors, help text, label, value and widget placement.
+
+## Usage
+
+```jsx
+import React from 'react';
+import Field from 'terra-form/lib/Fieldset';
```
Comment (windse7en) [inline packages/terra-form/docs/components/Field.md]: Is there any way we can get rid the `lib`,  just use `import Field from 'terra-form/Fieldset'`?

File: packages/terra-form/src/components/Field.jsx [hunk 17]
```diff
@@ -0,0 +1,66 @@
+/* eslint-disable react/jsx-boolean-value, jsx-a11y/label-has-for */
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import 'terra-base/lib/baseStyles';
```
Comment (windse7en) [inline packages/terra-form/src/components/Field.jsx]: A stupid question: Why do our components have to import `baseStyles`?

File: packages/terra-form/docs/components/Field.md (hunk 16 above)
Comment (JakeLaCombe) [inline packages/terra-form/docs/components/Field.md]: I envisioned having the following file structure for terra-form.

     - terra-form
        - lib
              - components
              - validations

It gives the component better file structure in my opinion. Do others have any views on this?

File: packages/terra-form/src/components/Field.jsx [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * Error message for when the input is invalid
+   */
+  error: PropTypes.node,
+  /**
+   * Help element to display with the input
+   */
+  help: PropTypes.node,
+  /**
+   * Determines whether the field in an inline field
```
Comment (mjhenkes) [inline packages/terra-form/src/components/Field.jsx]: "Determines whether the field is an inline field"

File: packages/terra-form/src/components/Fieldset.jsx [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * Help element to display with the input
+   */
+  help: PropTypes.node,
+  /**
+   * The value for the htmlFor property on the label
+   */
+  htmlFor: PropTypes.string,
+  /**
+   * Determines whether the field in an inline field
```
Comment (mjhenkes) [inline packages/terra-form/src/components/Fieldset.jsx]: "Determines whether the field is an inline field"

File: packages/terra-form/docs/components/Field.md (hunk 16 above)
Comment (mjhenkes) [inline packages/terra-form/docs/components/Field.md]: Do you anticipate having any additional folders at the lib level? I lean towards agreeing with @windse7en

File: packages/terra-form/src/components/Field.jsx (hunk 17 above)
Comment (mjhenkes) [inline packages/terra-form/src/components/Field.jsx]: @bjankord has a good explanation here: https://github.com/cerner/terra-core/pull/251

File: packages/terra-form/docs/components/Field.md (hunk 16 above)
Comment (JakeLaCombe) [inline packages/terra-form/docs/components/Field.md]: Those are the only ones I anticipate atm. Given that, I can move everything up to lib.

File: packages/terra-form/lib/_variables.scss (hunk 5 above)
Comment (JakeLaCombe) [inline packages/terra-form/lib/_variables.scss]: I didn't see form theming requirements, so I removed them.

https://github.com/cerner/terra-core/pull/352/commits/0943e9d4159fa1dd3f2e8325ede53d48c666d717

Comment (JakeLaCombe) [inline packages/terra-form/lib/_variables.scss]: Removed, as I couldn't find theming requirements. https://github.com/cerner/terra-core/pull/352/commits/12196782a827e85c45446500ecbf1c7eef08b4e6

File: packages/terra-form/src/components/Field.jsx (hunk 17 above)
Comment (windse7en) [inline packages/terra-form/src/components/Field.jsx]: Thanks.

File: packages/terra-form/docs/components/Field.md (hunk 16 above)
Comment (JakeLaCombe) [inline packages/terra-form/docs/components/Field.md]: Moved them over. https://github.com/cerner/terra-core/pull/352/commits/70bc8f0a6b07a540ff52743bd71d7063fba8fe5f

File: packages/terra-form/src/components/Fieldset.jsx (hunk 19 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Fieldset.jsx]: https://github.com/cerner/terra-core/pull/352/commits/d43575884989c3f9802166d288225744244ad996

File: packages/terra-form/src/components/Field.jsx (hunk 18 above)
Comment (JakeLaCombe) [inline packages/terra-form/src/components/Field.jsx]: https://github.com/cerner/terra-core/pull/352/commits/d43575884989c3f9802166d288225744244ad996

File: packages/terra-form/docs/Fieldset.md [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+Generic form fieldset component which handles the layout of a standard form fieldset including errors, help text, label, value and widget placement.
+
+## Usage
+
+```jsx
+import React from 'react';
+import Fieldset from 'terra-form/lib/Fieldset';
+
+<Fieldset
+  type="checkbox"
```
Comment (bjankord) [inline packages/terra-form/docs/Fieldset.md]: Is type a valid prop here? I didn't see it defined in the proptypes for `packages/terra-form/src/Fieldset.jsx`.

File: packages/terra-form/docs/Fieldset.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+  legend="Do you have any Children?"
+  name="children_present"
+  value="children_present"
```
Comment (bjankord) [inline packages/terra-form/docs/Fieldset.md]: Are `name` and `value` valid props here? I didn't see it defined in the proptypes for `packages/terra-form/src/Fieldset.jsx`.

Comment (JakeLaCombe) [inline packages/terra-form/docs/Fieldset.md]: Looks like I missed a few revisions from before. 

https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

File: packages/terra-form/docs/Fieldset.md (hunk 20 above)
Comment (JakeLaCombe) [inline packages/terra-form/docs/Fieldset.md]: This is outdated. Fixed here. https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

File: packages/terra-site/src/examples/form/examples/Input.jsx (hunk 12 above)
Comment (JakeLaCombe) [inline packages/terra-site/src/examples/form/examples/Input.jsx]: I decided to add it for this pull request. Setups the easy addition of future controlled inputs.

https://github.com/cerner/terra-core/pull/352/commits/33ffe7b0aba15bcf8a600b929fea2d7f7d0c3e30

File: packages/terra-form/src/Fieldset.jsx [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+  children: null,
+  error: null,
+  help: null,
+  htmlFor: null,
+  isInline: false,
+  legend: null,
+  required: false,
+};
+
+const Fieldset = ({ children, error, help, htmlFor, isInline, legend, required, ...customProps }) => {
```
Comment (nickpith) [inline packages/terra-form/src/Fieldset.jsx]: I don't see htmlFor being used anywhere in this component.

Comment (JakeLaCombe) [inline packages/terra-form/src/Fieldset.jsx]: It doesn't belong here. Removed.

https://github.com/cerner/terra-core/pull/352/commits/2c7f7c6030aea22430c5009637aca5681b94a9a6

**Comment by JakeLaCombe** [comment]
Added htmlFor to the field. https://github.com/cerner/terra-core/pull/352/commits/c7029abf4f24439a2f1691787ab33ecd471e8766

I can't find a good way to generate unique ids for the htmlFor attribute, so I left that for the consumer to pass in. It's not throwing warnings in the browser or lint though, so it should be ok to leave the default as undefined.

**Comment by mmalaker** [comment]
+1 functional review

**Comment by neilpfeiffer** [comment]
+1 for checkpoint UX Review

One thing we will want to look at is the how to better style fieldsets, particularly the helptext and errors which look a little distant when fields are inline and don't also have helptext or errors. I'll make a note of some variations we can try.

**Pull request merged**